fix(#479): render groups with no ref attribute

garethbowen opened the pull request:
Closes #479

Also includes a significant refactor to reduce the Group types from 3 (structural, presentation, logical) to 1. There was some inconsistency here with a) naming internally, and b) the spec. Another option would have been to retain two types (presentation and non-presentation) but as this is not useful right now I decided to go back to 1 type and we can add subtypes back when it becomes an issue.

### I have verified this PR works in these browsers (latest versions):

- [ ] Chrome
- [x] Firefox
- [ ] Safari (macOS)
- [ ] Safari (iOS)
- [ ] Chrome for Android
- [ ] Not applicable

### What else has been done to verify that this works as intended?

Using the form provided in the issue.

### Why is this the best possible solution? Were any other approaches considered?

1. Making a simpler change which just checked for either Group or Subtree, but in the end the flattening was preferred to clean up the inconsistencies.

### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

This is a breaking change in terms of the engine API, but front end users will not notice.

### Do we need any specific form for testing your changes? If so, please attach one.

The form attached in the issue is useful.

### What's changed

Group types have been reconciled so the web-forms UI handles groups with no ref in the same way as groups with a ref.

latin-panda commented on packages/web-forms/src/components/form-layout/FormGroup.vue:
```diff
@@ -1,10 +1,10 @@
 import QuestionList from './QuestionList.vue';
 
-const props = defineProps<{ node: GroupNode }>();
+const props = defineProps<{ node: GroupNode | SubtreeNode }>();
```
To cross-post the conversation from before. 

A Subtree always appears as either a logical group or a presentation group. It might be possible to simplify this by having a single Node representation for both, allowing the client to handle small rendering differences (such as when appearances is null) within the Group Vue component.

It was a bit confusing at first: What is a Subtree? What else can be a Subtree? Why are repeats not Subtree? How does "Subtree" align with the terminology found in ODK's spec documentation?
It took longer to feel confident about my interpretation being correct, and confidence about side effects if I were to modify SubTree in the future.

garethbowen replied:
Phew... well... this thread just keeps coming. Firstly some useful comments in the code:

https://github.com/getodk/web-forms/blob/30997b55d582f41945b90c214d6ddb770f440b20/packages/xforms-engine/src/parse/body/group/BaseGroupDefinition.ts#L14-L37

There are various comments sprinkled through the code like:

> // TODO: obviously there is a naming inconsistency emerging here.

1. I don't see any value in the "subtree" nomenclature. It's not used anywhere else that I can see. Can we find and replace "subtree" -> "group"?
2. web-forms talks about logical, presentation, and structural groups as being distinct, but the spec doesn't mention structural groups at all. The only difference between structural and presentation is the latter includes a label. I don't think this distinction is useful in web forms and we should reduce the choices to "logical" and "presentation" groups, where presentation groups may or may not have a label.
3. I do think there's value in the type system reinforcing the differences between logical and presentation groups. However it would be useful to have views deal with any group where the logic is the same (or similar). I propose introducing a parent type for the client api, ie: LogicalGroupNode and PresentationGroupNode would both extend  GroupNode. We should also update the internal engine types at the same time to be consistent with this naming.

garethbowen replied:
I don't think there's much difference in the UI aside from appearances. A few notes from [the spec](https://getodk.github.io/xforms-spec/#groups)...

- A logical group’s ref is used as the context node for the relative ref paths of its descendants.
- a logical group can also contain a relevant attribute on its <bind> element

But that's about it.

garethbowen replied:
A summary of our conclusion from the call just now...

1. Replace the term Subtree with Group as it's more obvious what it is.
2. Define types for PresentationGroup (a group which has a label child element) and NonPresentationGroup (is there a better name for this?) because there _should_ be a visual distinction between the two.
3. Make no type distinction for groups with or without a ref - the latter Node will have `appearances = null`.

@lognaturel @latin-panda Is this what you took out of the meeting too?

latin-panda replied:
Sounds good to me

Naming options for NonPresentationGroup:
- StructuralGroup: seemed like a natural choice to me
- LogicalGroup: aligns with specs — initially, someone might question what "logical" means here.

lognaturel replied:
A logical group is any group that has a `ref` attribute so it would not be accurate to use here! Super confusing, I know. `NonPresentationGroup` seems like the most clear to me.

latin-panda reviewed (commented):
I tested other presentation and logical group forms, as well as repeats. It works fine! 

I left a suggestion below. Let me know what you think!

latin-panda commented on packages/common/src/fixtures/groups/presentation-group.xml:
```diff
@@ -0,0 +1,79 @@
+    </model>
+  </h:head>
+  <h:body>
+    <group appearance="field-list">
```
Good question, yes, leaving just one should be fine. To render one means the FormQuestion component has been reached, and then any other question type can render as well.

garethbowen commented:
@latin-panda As discussed I've refactored the *Group to greatly simplify it and removing naming inconsistencies. After some investigation it was clear there is very little distinction between any of the three group types either in the UI or the engine so I decided to flatten down to just one. I removed most of the references to "subtree" where "group" was more descriptive but left references where it's referring to a subset of the document.

Let me know what you think!

lognaturel commented:
Haven't looked in detail but the simplification sounds good to me. I would expect tests at https://github.com/getodk/web-forms/blob/6cfff8b4c5a2cf6a23a71ef6d4308343bccd2436/packages/odk-web-forms/src/lib/xform/model/ModelDefinition.test.ts#L479 and https://github.com/getodk/web-forms/blob/6cfff8b4c5a2cf6a23a71ef6d4308343bccd2436/packages/odk-web-forms/src/lib/xform/model/ModelDefinition.test.ts#L576 to be addressed in some way, either in this PR or in follow-up. I haven't dug deeply but I think the first can be modified to pass and the second can be removed because there isn't a need for the structural group concept I think.

lognaturel commented on .changeset/great-deer-march.md:
```diff
@@ -0,0 +1,5 @@
+'@getodk/web-forms': minor
+---
+
+Render groups without a ref attribute
```
There's now also a change to the engine that would be helpful to document as a minor change (because we're pre-1.0)?

garethbowen commented on .changeset/great-deer-march.md:
```diff
@@ -0,0 +1,5 @@
+---
+'@getodk/web-forms': minor
```
```suggestion
'@getodk/xforms-engine': minor
'@getodk/web-forms': minor
```

lognaturel commented:
Appreciate you! Nice to have a bug fix and a nice simplification. 🎉
